### DISCUSSION TITLE
How can I mock h3.useSession?

### DISCUSSION THREAD

**Discussion by eddie4c7** [vitest-dev/vitest#5536] [Q&A] [answered]
Hello.

I wanted to mock h3.useSession by spyOn to verify called session.update.
But could not it with message "spyon could not find an object to spy upon".
How can I do it?

Code
```ts
import { test, vi } from 'vitest'
import h3 from 'h3'

test('valid user', async () => {
  const session = {
    id: '',
    data: {},
    update: vi.fn(),
    clear: vi.fn()
  }
  vi.spyOn(h3, 'useSession').mockResolvedValue(session)
})
```

Output
```
Error: spyOn could not find an object to spy upon
```

h3 handle session
https://h3.unjs.io/examples/handle-session

**Comment by hi-ogawa** [comment, marked as answer]
Their documentation says `useSession` is available from named exports `import { useSession } from "h3"`, which is usually different from default export's property `import h3 from "h3"; h3.useSession;`.

In this case, you might need to use module mocking [`vi.mock`](https://vitest.dev/api/vi.html#vi-mock), something like:

```ts
import { test, vi, expect } from "vitest";

const mockSession = vi.hoisted(() => ({
  id: '',
  data: {},
  update: vi.fn(),
  clear: vi.fn()
});

vi.mock("h3", async (importOriginal) => {
  return {
    ...await importOriginal(),
    useSession: async () => mockSession,
  }
})

test("...", () => {
  // ... do something with h3 ...
  
  // assert calls
  expect(mockSession.update).toHaveBeenCalled()
})
```

**Reply by eddie4c7** [reply]
I could do it by this way, thanks!
